C builderr with interfaces that have option fields and embedded structs that implement the interface

### Describe the bug

An interface that has an option field generates a C error when a struct which embeds another struct that implements the interface tries to access the option field.

### Reproduction Steps

```v
import time

interface InterfaceObject {
mut:
	duration ?time.Duration
	update()
}

struct Object {
mut:
	duration ?time.Duration
}

fn (mut obj Object) update() {
	duration := obj.duration or { time.second }
	println(duration)
}

struct FooObject {
	Object
}

fn (mut obj FooObject) update() {
	duration := obj.duration or { time.millisecond * 500 }
	println(duration)
}

fn main() {
	mut object := InterfaceObject(FooObject{})
	println(object)
}

```

### Expected Behavior

Expected code to compile and print the object.

### Current Behavior

C error:
```
C:/Users/imado/AppData/Local/Temp/v_0/test.16517445185759149149.tmp.c:519: warning: WINVER redefined
C:/Users/imado/AppData/Local/Temp/v_0/test.16517445185759149149.tmp.c:7361: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/imado/AppData/Local/Temp/v_0/test.16517445185759149149.tmp.c:15044: warning: assignment from incompatible pointer type
C:/Users/imado/AppData/Local/Temp/v_0/test.16517445185759149149.tmp.c:15050: warning: assignment from incompatible pointer type
C:/Users/imado/AppData/Local/Temp/v_0/test.16517445185759149149.tmp.c:15056: warning: assignment from incompatible pointer type
C:/Users/imado/AppData/Local/Temp/v_0/test.16517445185759149149.tmp.c:15064: warning: assignment from incompatible pointer type
C:/Users/imado/AppData/Local/Temp/v_0/test.16517445185759149149.tmp.c:15090: warning: assignment from incompatible pointer type
C:/Users/imado/AppData/Local/Temp/v_0/test.16517445185759149149.tmp.c:15090: warning: assignment from incompatible pointer type
C:/Users/imado/AppData/Local/Temp/v_0/test.16517445185759149149.tmp.c:15110: warning: assignment from incompatible pointer type
C:/Users/imado/AppData/Local/Temp/v_0/test.16517445185759149149.tmp.c:15112: warning: assignment from incompatible pointer type
C:/Users/imado/AppData/Local/Temp/v_0/test.16517445185759149149.tmp.c:15112: warning: assignment from incompatible pointer type
C:/Users/imado/AppData/Local/Temp/v_0/test.16517445185759149149.tmp.c:15132: warning: assignment from incompatible pointer type
C:/Users/imado/AppData/Local/Temp/v_0/test.16517445185759149149.tmp.c:15313: error: field not found: duration
builder error: 
==================
C error. This should never happen.
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V full version: V 0.4.1 68cbf27.3fab321

### Environment details (OS name and version, etc.)

```

OS: windows, Microsoft Windows 11 Pro v22621 64-bit
Processor: 16 cpus, 64bit, little endian, 

getwd: C:\Users\imado\Documents\cyberian_tiger
vexe: C:\Users\imado\v\v.exe
vexe mtime: 2023-09-23 02:25:12

vroot: OK, value: C:\Users\imado\v
VMODULES: OK, value: C:\Users\imado\.vmodules
VTMP: OK, value: C:\Users\imado\AppData\Local\Temp\v_0

Git version: git version 2.33.1.windows.1
Git vroot status: weekly.2023.38-24-g3fab3213
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,
operable program or batch file.

thirdparty/tcc status: thirdparty-windows-amd64 e90c2620
```
> [!IMPORTANT]
> You can vote for this issue using the 👍 reaction. More votes increase the issue's priority
> for developers.
>
> Take into account that only the 👍 reaction counts as a vote.
> Only reactions to the issue itself will be counted as votes, not comments.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -3633,7 +3633,22 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 			g.write('*(')
 		}
 		g.expr(node.expr)
-		if node.expr_type.is_ptr() {
+		for i, embed in node.from_embed_types {
+			embed_sym := g.table.sym(embed)
+			embed_name := embed_sym.embed_name()
+			is_left_ptr := if i == 0 {
+				node.expr_type.is_ptr()
+			} else {
+				node.from_embed_types[i - 1].is_ptr()
+			}
+			if is_left_ptr {
+				g.write('->')
+			} else {
+				g.write('.')
+			}
+			g.write(embed_name)
+		}
+		if node.expr_type.is_ptr() && node.from_embed_types.len == 0 {
 			g.write('->')
 		} else {
 			g.write('.')

--- a/vlib/v/tests/interface_with_option_field_test.v
+++ b/vlib/v/tests/interface_with_option_field_test.v
@@ -1,0 +1,32 @@
+import time
+
+interface InterfaceObject {
+mut:
+	duration ?time.Duration
+	update()
+}
+
+struct Object {
+mut:
+	duration ?time.Duration
+}
+
+fn (mut obj Object) update() {
+	duration := obj.duration or { time.second }
+	println(duration)
+}
+
+struct FooObject {
+	Object
+}
+
+fn (mut obj FooObject) update() {
+	duration := obj.duration or { time.millisecond * 500 }
+	println(duration)
+}
+
+fn test_interface_with_option_field() {
+	mut object := InterfaceObject(FooObject{})
+	println(object)
+	assert '${object.duration}' == 'Option(none)'
+}

